keps/sig-node/5224-node-resource-discovery/README.md

```diff
@@ -0,0 +1,1211 @@
+    resources:
+      - name: "cpu"
+        capacity: 1
+  - name: "core-2"
+    type: "core"
+    parent: "socket-1"
+    resources:
+      - name: "cpu"
+        capacity: 1
+  - name: "socket-2"
+    type: "socket"
+    parent: "root"
+...
+```
+
+### kubelet
+
+The kubelet is updated to query the node resources via the new
+GetDynamicRuntimeConfig CRI endpoint. The [cAdvisor][cadvisor] MachineInfo is
+fully replaced with the information received from the CRI. The node capacity is
+queried at kubelet startup and the Node object on apiserver is only updated at
+this time. A separate enhancement proposal ([KEP-3953][kep-3953]) concentrates
+on runtime dynamic updates of resources.
+
+The kubelet keeps the streaming channel open and logs dynamic changes. However,
+the kubelet does not otherwise react to the dynamic changes after startup –
+dynamic node resize will be covered by [KEP-3953][kep-3953]. Node events about
+capacity changes will be generated when dynamic node resize
+([KEP-3953][kep-3953]) is enabled.
+
+The following diagram illustrates the kubelet resource discovery flow:
+
+```mermaid
+sequenceDiagram
+    participant k as kubelet
+    participant r as runtime
+    actor c as system/controller
+    r->c: detect resources
+    k->>r: open streaming interface
+    r->>k: send available resources
+    k->>k: update the node object, finish startup
+    Note over k, r: at startup
+    c->>r: resources change
+    r->>k: send updated resources to the kubelet
+    opt NodeResourceHotPlug=true
+        k->>k: re-initialize managers, update the node object
+    end
+```
+
+The handling of system-reserved and kube-reserved resources is not changed by
+this proposal.
+
+There is a fallback to cadvisor-based resource discovery even when the feature
+gate is enabled: if the CRI runtime does not support the new
+DynamicRuntimeConfigRequest rpc, the kubelet will resort to cadvisor. This will
+ensure compatibility with older versions of CRI runtimes.
+
+If the kubelet is not able to parse or consume the resource topology tree it
+will set the node into NotReady state, with node status, events (and kubelet
+logs) indicating the details.
+
+If the cpu manager static policy is active, the kubelet requires details of the
+hardware topology. It recognizes the well-known zone types (“NUMANode”,
+“Package”, “Core”, “CacheGroup”) to construct the HW topology to initialize the
+cpu manager, memory manager and topology manager. The cpu manager static policy
+requires information about CPU IDs and thus the kubelet requires the “cpu-ids”
+attribute to be present for zones containing cpu resources. If the cpu manager
+static policy is active and cpu-ids are not available the node is set into
+NotReady state. If the “none” policy is in use, cpu-ids are not required.
+
+The kubelet reads MacineID, BootID and SystemUUID from the attributes of the
+resource topology tree. If this information is not present the kubelet uses the
+cAdvisor MachineInfo as a fallback.
```

Comment:
forever? or will it stop after GA? What if this info is never available? should the kubelet exit?

Reply:
oops I should read the section above: set the node not ready for the last question

Reply:
I think this is a good question in general, whether the MachineID, BootID and SystemUUID should come from the runtime or should the kubelet figure out these itself. I put them here to be able to ditch cAdvisor MachineInfo completely. Glad to hear opinions on this.